src/Compilers/Extension/CompilerPackage.cs

```diff
@@ -4,25 +4,30 @@
 using System.Collections.Generic;
 using System.IO;
 using System.Reflection;
+using System.Threading;
 using EnvDTE;
 using Microsoft.VisualStudio.Shell;
 using Microsoft.VisualStudio.Shell.Interop;
+using Task = System.Threading.Tasks.Task;
 
 namespace Roslyn.Compilers.Extension
 {
-    [ProvideAutoLoad(UIContextGuids.SolutionExists)]
-    public sealed class CompilerPackage : Package
+    [ProvideAutoLoad(UIContextGuids.SolutionExists, PackageAutoLoadFlags.BackgroundLoad)]
+    [PackageRegistration(UseManagedResourcesOnly = true, AllowsBackgroundLoading = true)]
+    public sealed class CompilerPackage : AsyncPackage
     {
         public static string RoslynHive = null;
         
-        protected override void Initialize()
+        protected override async Task InitializeAsync(CancellationToken cancellationToken, IProgress<ServiceProgressData> progress)
         {
-            base.Initialize();
+            await base.InitializeAsync(cancellationToken, progress).ConfigureAwait(true);
+
+            await JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
 
             var packagePath = Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location);
 
             string localRegistryRoot;
-            var reg = (ILocalRegistry2)this.GetService(typeof(SLocalRegistry));
+            var reg = (ILocalRegistry2)await GetServiceAsync(typeof(SLocalRegistry)).ConfigureAwait(true);
             reg.GetLocalRegistryRoot(out localRegistryRoot);
             var registryParts = localRegistryRoot.Split('\\');
```

Comment:
hmmm.. looking at MSBUILD, that seems thread-safe as well.

http://index/?query=Microsoft.Build.Evaluation.ProjectCollection.GlobalProjectCollection.DisableMarkDirty&rightProject=Microsoft.Build&file=Definition%5CProjectCollection.cs&line=748

what is the thing that require it to run on UI thread?

Reply:
underneath, people are saying GetserviceAsync can return null, so probably null check needed here?

Reply:
> underneath, people are saying GetserviceAsync can return null, so probably null check needed here?

It sounded like they were saying it can return null if cancellation occurs, so I'm adding calls to `ThrowIfCancellationRequested` instead. If it specifically needs null checks, let me know.

Reply:
You should *always* perform a null check after a GetService or GetServiceAsync call. There are half a dozen reasons (most of them in legit conditions) for it to return null. If you require it to be non-null, use `Assumes.Present(reg)` so that it throws a meaningful exception with a useful message if it's null. Otherwise, you stand to throw NRE sometime later, which is often more ambiguous and therefore less actionable.